File: web/src/ChatSession.ts
import EdgeVeda from './index';
import { ChatMessage, ChatRole, SystemPromptPreset, getSystemPrompt } from './ChatTypes';
import { ChatTemplate, formatMessages } from './ChatTemplate';

/**
 * Manages a multi-turn chat conversation with context tracking
 */
export class ChatSession {
  private edgeVeda: EdgeVeda;
  private messages: ChatMessage[] = [];
  private maxContextLength: number;
  private template: ChatTemplate;
  private systemPromptText: string | null;
  
  /**
   * Creates a new chat session
   */
  constructor(
    edgeVeda: EdgeVeda,
    options: {
      systemPrompt?: SystemPromptPreset;
      maxContextLength?: number;
      template?: ChatTemplate;
    } = {}
  ) {
    this.edgeVeda = edgeVeda;
    this.maxContextLength = options.maxContextLength ?? 2048;
    this.template = options.template ?? ChatTemplate.LLAMA3;
    
    // Add system prompt if provided
    this.systemPromptText = options.systemPrompt ? getSystemPrompt(options.systemPrompt) : null;
    if (this.systemPromptText) {
      this.messages.push({
        role: ChatRole.SYSTEM,
        content: this.systemPromptText,
        timestamp: new Date(),
      });
    }
  }
  
  /**
   * Sends a message and returns the complete response
   */
  async send(message: string, options?: any): Promise<string> {
    // Add user message
    this.messages.push({
      role: ChatRole.USER,
      content: message,
      timestamp: new Date(),
    });
    
    // Format the prompt
    const prompt = this.formatPrompt();
    
    // Generate response
    const response = await this.edgeVeda.generate({ prompt, ...options });
    
    // Add assistant response
    this.messages.push({
      role: ChatRole.ASSISTANT,
      content: response,
      timestamp: new Date(),
    });
    
    return response;
  }
  
  /**
   * Sends a message and streams the response token by token
   */
  async *sendStream(message: string, options?: any): AsyncGenerator<string, void, unknown> {
    // Add user message
    this.messages.push({
      role: ChatRole.USER,
      content: message,
      timestamp: new Date(),
    });
    
    // Format the prompt
    const prompt = this.formatPrompt();
    let fullResponse = '';
    
    // Stream tokens
    for await (const chunk of this.edgeVeda.generateStream({ prompt, ...options })) {
      fullResponse += chunk.token;
      yield chunk.token;
    }
    
    // Add complete assistant response
    this.messages.push({
      role: ChatRole.ASSISTANT,
      content: fullResponse,
      timestamp: new Date(),
    });
  }
  
  /**
   * Resets the conversation, keeping only the system prompt
   */
  async reset(): Promise<void> {
    const systemMsg = this.messages.find(m => m.role === ChatRole.SYSTEM);
    this.messages = systemMsg ? [systemMsg] : [];
    await this.edgeVeda.resetContext();
  }
  
  /**
   * Returns the number of conversation turns (user messages)
   */
  get turnCount(): number {
    return this.messages.filter(m => m.role === ChatRole.USER).length;
  }
  
  /**
   * Returns the estimated context usage as a percentage (0.0 to 1.0)
   */
  get contextUsage(): number {
    const totalTokens = this.estimateTokens();
    return totalTokens / this.maxContextLength;
  }
  
  /**
   * Returns all messages in the conversation
   */
  get allMessages(): ChatMessage[] {
    return [...this.messages];
  }
  
  /**
   * Returns the last N messages in the conversation
   */
  lastMessages(count: number): ChatMessage[] {
    const startIndex = Math.max(0, this.messages.length - count);
    return this.messages.slice(startIndex);
  }
  
  /**
   * Formats all messages into a prompt using the chat template
   */
  private formatPrompt(): string {
    return formatMessages(this.template, this.messages);
  }
  
  /**
   * Estimates the total token count for all messages
   * Uses a rough heuristic: ~4 characters per token
   */
  private estimateTokens(): number {
    const totalChars = this.messages.reduce((sum, msg) => sum + msg.content.length, 0);
    return Math.floor(totalChars / 4);
  }
}

File: swift/Sources/EdgeVeda/ChatSession.swift
import Foundation

/// Manages a multi-turn chat conversation with context tracking
@available(iOS 15.0, macOS 12.0, *)
public actor ChatSession {
    // MARK: - Properties
    
    /// The EdgeVeda instance used for generation
    private let edgeVeda: EdgeVeda
    
    /// The conversation message history
    private var messages: [ChatMessage] = []
    
    /// Maximum context length in tokens
    private let maxContextLength: Int
    
    /// The system prompt (if any)
    private let systemPrompt: String?
    
    /// The chat template used for formatting
    private let template: ChatTemplate
    
    // MARK: - Initialization
    
    /// Creates a new chat session
    /// - Parameters:
    ///   - edgeVeda: The EdgeVeda instance to use for generation
    ///   - systemPrompt: The system prompt preset to use (defaults to .assistant)
    ///   - maxContextLength: Maximum context length in tokens (defaults to 2048)
    ///   - template: The chat template to use (defaults to .llama3)
    public init(
        edgeVeda: EdgeVeda,
        systemPrompt: SystemPromptPreset = .assistant,
        maxContextLength: Int = 2048,
        template: ChatTemplate = .llama3
    ) {
        self.edgeVeda = edgeVeda
        self.maxContextLength = maxContextLength
        self.template = template
        
        // Add system prompt if provided
        self.systemPrompt = systemPrompt.text
        if let prompt = self.systemPrompt {
            self.messages.append(ChatMessage(role: .system, content: prompt))
        }
    }
    
    // MARK: - Public Methods
    
    /// Sends a message and returns the complete response
    /// - Parameters:
    ///   - message: The user message to send
    ///   - options: Generation options
    /// - Returns: The assistant's response
    public func send(_ message: String, options: GenerateOptions = GenerateOptions()) async throws -> String {
        // Add user message
        messages.append(ChatMessage(role: .user, content: message))
        
        // Format the prompt
        let prompt = formatPrompt()
        
        // Generate response
        let response = try await edgeVeda.generate(prompt, options: options)
        
        // Add assistant response
        messages.append(ChatMessage(role: .assistant, content: response))
        
        return response
    }
    
    /// Sends a message and streams the response token by token
    /// - Parameters:
    ///   - message: The user message to send
    ///   - options: Generation options
    /// - Returns: An async stream of response tokens
    public func sendStream(_ message: String, options: GenerateOptions = GenerateOptions()) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            Task {
                // Add user message
                await self.addMessage(ChatMessage(role: .user, content: message))
                
                let prompt = await self.formatPrompt()
                var fullResponse = ""
                
                do {
                    for try await token in await self.edgeVeda.generateStream(prompt, options: options) {
                        fullResponse += token
                        continuation.yield(token)
                    }
                    
                    // Add complete assistant response
                    await self.addMessage(ChatMessage(role: .assistant, content: fullResponse))
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }
    
    /// Resets the conversation, keeping only the system prompt
    public func reset() async throws {
        messages.removeAll()
        
        // Re-add system prompt if it exists
        if let prompt = systemPrompt {
            messages.append(ChatMessage(role: .system, content: prompt))
        }
        
        // Reset the EdgeVeda context
        try await edgeVeda.resetContext()
    }
    
    /// Returns the number of conversation turns (user messages)
    public var turnCount: Int {
        messages.filter { $0.role == .user }.count
    }
    
    /// Returns the estimated context usage as a percentage (0.0 to 1.0)
    public var contextUsage: Double {
        let totalTokens = estimateTokens()
        return Double(totalTokens) / Double(maxContextLength)
    }
    
    /// Returns all messages in the conversation
    public var allMessages: [ChatMessage] {
        messages
    }
    
    /// Returns the last N messages in the conversation
    /// - Parameter count: Number of messages to return
    /// - Returns: The last N messages
    public func lastMessages(count: Int) -> [ChatMessage] {
        let startIndex = max(0, messages.count - count)
        return Array(messages[startIndex...])
    }
    
    // MARK: - Private Methods
    
    /// Adds a message to the conversation history
    private func addMessage(_ message: ChatMessage) {
        messages.append(message)
    }
    
    /// Formats all messages into a prompt using the chat template
    private func formatPrompt() -> String {
        template.format(messages: messages)
    }
    
    /// Estimates the total token count for all messages
    /// Uses a rough heuristic: ~4 characters per token
    private func estimateTokens() -> Int {
        let totalChars = messages.reduce(0) { $0 + $1.content.count }
        return totalChars / 4
    }
}